src/SDK/Trace/Sampler/AttributeBasedSampler.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\SDK\Trace\Sampler;
+
+use InvalidArgumentException;
+use OpenTelemetry\API\Behavior\LogsMessagesTrait;
+use OpenTelemetry\Context\ContextInterface;
+use OpenTelemetry\SDK\Common\Attribute\AttributesInterface;
+use OpenTelemetry\SDK\Trace\SamplerInterface;
+use OpenTelemetry\SDK\Trace\SamplingResult;
+
+/**
+ * @phan-file-suppress PhanParamTooFewUnpack
+ *
+ * Attribute-based sampler. This sampler should be called after a ParentBased sampler, and requires
+ * another sampler (such as AlwaysOnSampler).
+ * For example, ParentBased,AttributeBased(<deny if url.path attribute matches /health>),TradeIdRatio(0.5) means
+ * 1. Trace if parent is already sampled; otherwise
+ * 2. Do not sample if url.path attribute is set and matches /health; otherwise
+ * 3. Sample 50% of traces
+ *
+ * Two modes exist:
+ * "allow" mode: always sample if attribute exists and matches the pattern, else do not sample
+ * "deny" mode: do not sample if attribute exists and matches the pattern, else defer to next sampler
+ */
+class AttributeBasedSampler implements SamplerInterface
+{
+    use LogsMessagesTrait;
+
+    public const ALLOW = 'allow';
+    public const DENY = 'deny';
+    public const MODES = [
+        self::ALLOW,
+        self::DENY,
+    ];
+
+    private SamplerInterface $delegate;
+    private string $mode;
+    private string $attribute;
+    private string $pattern;
+
+    /**
+     * @param SamplerInterface $delegate The sampler to defer to if a decision is not made by this sampler
+     * @param string $mode Sampling mode (deny or allow)
+     * @param string $attribute The SemConv trace attribute to test against, eg http.path, http.method
+     * @param string $pattern The PCRE regex pattern to match against, eg /\/health$|\/test$/
+     */
+    public function __construct(SamplerInterface $delegate, string $mode, string $attribute, string $pattern)
+    {
+        if (!in_array($mode, self::MODES)) {
+            throw new InvalidArgumentException('Unknown Attribute sampler mode: ' . $mode);
+        }
+        $this->delegate = $delegate;
+        $this->mode = $mode;
+        $this->attribute = $attribute;
+        $this->pattern = $pattern;
+    }
+
+    public function shouldSample(ContextInterface $parentContext, string $traceId, string $spanName, int $spanKind, AttributesInterface $attributes, array $links): SamplingResult
+    {
+        switch ($this->mode) {
+            case self::ALLOW:
+                if (!$attributes->has($this->attribute)) {
+                    return new SamplingResult(SamplingResult::DROP);
```

Comment:
> I did some prototyping that unifies Javas RuleBasedRoutingSampler and LinksBasedSampler

I was just reading about the idea of a rule-based sampler in [this OTEP](https://github.com/open-telemetry/oteps/pull/240/files#diff-e53a876f51bdcf6825797cf103af52c760750d8288fc031b10222ab238ff2ecaR121) and I like the idea of using that approach over what I've done here, given the increased usefulness (and hopefully also compatibility with the proposed sampling configuration).

The prototype looks great, I think we should go with that approach. For the time being, it might not be configurable via env vars (and so not available to auto-instrumentation/sdk-autoloading) due to the complexity of the configuration.